### PR TITLE
Migrate drawer math eval from meval to exmex (#64, #28)

### LINKED ISSUE #28: Track meval dependency on deprecated nom 1.x
## Context

The `meval` crate (math expression evaluator, added in PR #27) depends on `nom v1.2.4` which triggers a future-incompatibility warning:

```
warning: the following packages contain code that will be rejected by a future version of Rust: nom v1.2.4
```

## Current status

- **Not blocking** — works fine on current Rust (1.94.0)
- meval itself is stable and complete (1.2M downloads, last release 2018)
- The warning is informational — Rust has not set a date for breaking nom 1.x

## Future options (when/if Rust breaks nom 1.x)

1. **Local fork** — copy meval source into our workspace, update nom to 7.x (requires parser rewrite due to nom API changes)
2. **Contribute upstream** — PR to rekka/meval-rs (repo inactive since 2018, PR may not be reviewed)
3. **Alternative crate** — evaluate `fasteval`, `mexprp`, or other maintained math evaluators

## Action

No action needed today. Revisit when Rust deprecation becomes a compiler error.

### LINKED ISSUE #64: Replace meval with exmex to eliminate nom 1.2.4 future-incompat warning
## Background

Our build emits this warning on every compile:

```
warning: the following packages contain code that will be rejected by a future version of Rust: nom v1.2.4
```

Tracked originally in #28. The culprit is `meval 0.2.0 → nom 1.2.4`. `meval` was last released in 2018 and hasn't been updated in ~8 years, so there's no upstream fix coming. To eliminate the warning permanently, we need to migrate off `meval`.

## Current usage

Single file: `crates/nwg-drawer/src/ui/math.rs` — inline calculator in the drawer search bar.

```rust
match meval::eval_str(trimmed) {
    Ok(val) if val.is_nan() => MathResult::Error("undefined".to_string()),
    Ok(val) if val.is_infinite() => MathResult::Error("overflow".to_string()),
    Ok(val) => MathResult::Value(val),
    Err(_) => MathResult::NotMath,
}
```

Functional requirements (non-negotiable — all must be preserved):
- Basic arithmetic: `+`, `-`, `*`, `/`, `^`, `%`, unary minus, parentheses
- Decimals / floating-point
- Functions: `sin`, `cos`, `tan`, `sqrt`, `abs`, `ln`, `log`
- Constants: `pi`, `e`
- `fn eval(&str) -> Result<f64, Error>` shape API
- Must distinguish parse errors ("not math") from runtime errors (div by zero, NaN, overflow)

## Research summary

Evaluated 7 alternatives (late 2026 data):

| Crate | Last release | License | Maintained | Verdict |
|---|---|---|---|---|
| **exmex** | 2025-10-14 | MIT/Apache | ✅ | **Winner** |
| evalexpr | 2025-11-26 | AGPL-3.0 (since v12) | ✅ | ❌ Viral license |
| fasteval | 2020-01-25 | MIT | ❌ 6 years | ❌ Same trap as meval |
| kalk | 2025-09-05 | MIT | ✅ | ❌ wasm-bindgen + GMP deps |
| rink-core | 2024-04-21 | MPL-2.0 | ✅ | ❌ Massive overkill |
| rsc | 2024-03-31 | MIT | ⚠️ slowing | Close second |
| meval-rs / meval | 2018 | MIT | ❌ | N/A (what we're leaving) |

## Chosen path: `exmex`

**Why**:
- Actively maintained, permissively licensed (MIT OR Apache-2.0)
- Nearly drop-in API: `exmex::eval_str::<f64>("2+2")`
- Covers all required operators/functions/constants out of the box
- Plus bonus functions we didn't have: `tanh`, `cbrt`, `floor`, `ceil`, `signum`, constants `pi`/`π`/`e`/`tau`
- No ancient deps — kills the nom 1.2.4 warning

**One tradeoff**:
- Pulls `regex` as a transitive dep (~200KB). Acceptable for our use case.

## Proposed implementation plan

1. Add `exmex` to workspace dependencies (pin to `^0.20`)
2. Remove `meval` from workspace dependencies
3. Update `crates/nwg-drawer/src/ui/math.rs`:
   - Replace `meval::eval_str(...)` with `exmex::eval_str::<f64>(...)`
   - Adjust error matching — exmex uses a single `ExError` struct; we can treat any `Err` as `NotMath` (same as we do now)
   - NaN/infinity handling stays identical (we check the `f64` result)
4. Run existing math test suite (18 tests in `math.rs`) — these will catch any behavioral drift
5. Smoke test the drawer search bar: type `2+2`, `sqrt(16)`, `sin(pi)`, `10/0`, `sqrt(-1)`, `2^10`, `pi`, `e`, malformed input
6. Verify the nom 1.2.4 warning is gone from `cargo build` output
7. Update the "Deviations from Go originals" section in README (current note says we use a custom arithmetic parser — needs correction since we're using exmex now)

## Behavioral differences to verify

- [ ] `sqrt(-1)` — exmex returns NaN, meval returns NaN. Both trigger our "undefined" error ✓
- [ ] `2^1024` — exmex returns Infinity, meval returns Infinity. Both trigger our "overflow" error ✓
- [ ] Incomplete expression like `"2+"` — should be `NotMath`, not an error value
- [ ] Parentheses balance: `"(1+2"` and `"1+2)"` → `NotMath`
- [ ] Function arity: `sqrt()` with no arg, `sin(1,2)` with too many → `NotMath`
- [ ] Whitespace handling: `"  2 + 2  "` should still work

## Risk level

**Low.** Single file change, ~10 lines of eval logic, strong existing test coverage. Worst case is a behavioral drift in an edge case that our tests catch.

## Related

- #28 — original nom 1.2.4 tracking issue (will be closed by this)

### PR DESCRIPTION
**Pull request by jasonherald**
Closes #64. Closes #28.

## Summary
- Swap `meval 0.2` → `exmex 0.20` in the workspace. Kills the `nom v1.2.4` future-incompat warning permanently — the unmaintained transitive dep is gone.
- Port the one caller (`crates/nwg-drawer/src/ui/math.rs`) to `FlatEx` with a small `DrawerOpsFactory` that preserves meval's user-facing behaviors exmex doesn't ship:
  - `pi` as an alias for `π`
  - `%` binary modulo at priority 3 (keeps `10 % 3 * 2 = 2` — exmex resolves same-priority right-to-left and its `*` is priority 2)
  - `log` rebound to base-10 (exmex's default `log` is natural log; calculators and meval treat `log` as log10)
- Free identifiers map to `NotMath` so typing app names in the search bar doesn't surface a math row.
- Add 6 regression tests covering each migration-specific behavior so any future trim of the ops factory fails loudly.
- README's "Math evaluation" deviation note updated to describe the actual current setup (was claiming a custom parser we never had).

## Test plan
- [x] `cargo test --workspace` — 287 unit tests pass (26 in `ui::math`, up from 20; 6 new regressions cover `pi` alias, `%` precedence, log10, free-identifier rejection, whitespace, `sin(pi)` display)
- [x] `cargo clippy --all-targets -- -D warnings` clean
- [x] `cargo deny check` — advisories / bans / licenses / sources OK
- [x] `cargo build --release` no longer emits the `nom v1.2.4` future-incompat warning
- [x] SonarQube quality gate OK (0 new violations)
- [x] Smoke tested locally — drawer math eval working as expected
- [ ] Reviewer to confirm against their daily drawer use

## Behavioral parity checklist (from #64)
- [x] `2+2`, `sqrt(16)`, `sin(pi)` — value results
- [x] `10/0` — "division by zero" maps to `overflow` (infinity) error
- [x] `sqrt(-1)` — `undefined` (NaN) error
- [x] `2^10` — value result; `2^1024` — `overflow` error
- [x] `pi`, `π`, `e` — constants
- [x] `2+`, `(3*`, `sqrt(` — `NotMath` (incomplete while typing)
- [x] `  2 + 2  ` — whitespace trimming
- [x] `10 % 3`, `10 % 3 * 2` — modulo behavior matches meval
- [x] `log(100) = 2` — base-10 log preserved

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added support for mathematical constants pi/π, modulo (%) operator, and base‑10 log in inline math evaluation.

* **Bug Fixes**
  * Improved parsing and error handling: unbound identifiers are treated as non-math; NaN and infinity map to clearer "undefined"/"overflow" results; whitespace handling more forgiving.

* **Documentation**
  * Updated documentation to list supported functions/constants and behaviors.

* **Tests**
  * Added regression tests covering constants, modulo precedence, chaining, and common functions.

### CODE REVIEW

File: crates/nwg-drawer/src/ui/math.rs [hunk 1]
```diff
@@ -12,16 +13,68 @@ pub enum MathResult {
...
+/// - `%` as a binary modulo operator — exmex documents `%` as a custom
+///   operator example but doesn't register it by default.
+/// - `log` redefined to base-10 — exmex's default `log` is the natural
+///   logarithm (same as `ln`), whereas calculators and meval treat `log`
+///   as log-base-10. `ln` stays available for natural log.
+///
+/// `%` priority matches exmex's `/` (prio 3), which is higher than `*`
+/// (prio 2). Within a single priority class exmex resolves
+/// right-associatively, so picking `/`'s priority keeps `10 % 3 * 2`
+/// evaluating as `(10 % 3) * 2 = 2` rather than `10 % (3 * 2) = 4`.
```
[the remark this answers is not in the document]
Comment (jasonherald) [reply crates/nwg-drawer/src/ui/math.rs]: Good catches, both addressed in 0486df2:

- **Doc comment**: the "right-associativity within a single priority class" framing was wrong — `%` (prio 3) and `*` (prio 2) aren't in the same priority class, so that's not what's keeping `10 % 3 * 2` at 2. Priority alone does it. Rewrote the block to describe the actual mechanic and cover the `/` same-priority case separately.
- **Test comment**: "priority 2, matching * and /" was stale — draft was prio 2, ship was prio 3. Updated to match the code.
- **Same-priority test**: added `assert_eq!(eval_val("19 % 5 / 2"), 2.0);` as suggested. This pins left-to-right evaluation at priority 3; if someone ever bumps `%`'s priority past `/`, that assertion flips to 1.5 and fails loudly.
